streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +52,222 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
+                    log.error("Failed to close restoring task {} due to the following error:", id, e);
+                    if (exception == null) {
+                        exception = e;
+                    }
+                } finally {
+                    revokedTaskIterator.remove();
+                    restoring.remove(id);
+                    for (final TopicPartition tp : task.partitions()) {
+                        restoredPartitions.remove(tp);
+                        restoringByPartition.remove(tp);
+                    }
+                    revokedChangelogs.addAll(task.changelogPartitions());
+                }
+            } else {
+                // All non-restoring tasks should have been removed from revokedRestoringTasks, so log an error if it's not in restoring
+                log.error("Task was revoked but cannot be found in the assignment");
+                exception = new IllegalStateException("Revoked a task that was not in the assignment");
+                revokedTaskIterator.remove();
+            }
+        }
 
         return exception;
     }
 
+    /**
+     * @throws TaskMigratedException if the task producer got fenced (EOS only)
+     */
+    boolean maybeResumeSuspendedTask(final TaskId taskId, final Set<TopicPartition> partitions, final List<TopicPartition> changelogs) {
+        if (suspended.containsKey(taskId)) {
+            final StreamTask task = suspended.get(taskId);
+            log.trace("Found suspended {} {}", taskTypeName, taskId);
+            suspended.remove(taskId);
+
+            if (task.partitions().equals(partitions)) {
+                task.resume();
+                try {
+                    transitionToRunning(task);
+                } catch (final TaskMigratedException e) {
+                    // we need to catch migration exception internally since this function
+                    // is triggered in the rebalance callback
+                    log.info("Failed to resume {} {} since it got migrated to another thread already. " +
+                        "Closing it as zombie before triggering a new rebalance.", taskTypeName, task.id());
+                    final RuntimeException fatalException = closeZombieTask(task);
+                    running.remove(taskId);
+                    runningByPartition.keySet().removeAll(task.partitions());
+                    if (fatalException != null) {
+                        throw fatalException;
+                    }
+                    throw e;
+                }
+                log.trace("Resuming suspended {} {}", taskTypeName, task.id());
+                return true;
+            } else {
+                log.warn("Couldn't resume task {} assigned partitions {}, task partitions {}", taskId, partitions, task.partitions());
+                suspended.remove(task.id());
+                running.remove(task.id());
+                runningByPartition.keySet().removeAll(task.partitions());
```

Comment:
oh, also here?

Reply:
We do add the changelogs, below -- or is that not what you're referring to?

Reply:
Nvm, got your point and fixed